### PR TITLE
V2 part 2: Flash-BEATs: SDPA attention + batched fbank (stacked on #191)

### PR DESCRIPTION
**Pull request by nkundiushuti**
Net-new Flash-BEATs work, rebuilt cleanly on top of #191 (assumes #191 merges first).
This branch currently works to load naturelm 1.1 and 1.5 on esp_research (tested).

## Contents
- **#160 Flash-BEATs** — replaces manual attention with `F.scaled_dot_product_attention` (Flash Attention 2 / mem-efficient dispatch) + batched mel filterbank (`_BatchedFbank`).
- **#174 David fix2** — remaining net-new bits (old sdpa test removed/renamed).
- **Pre-existing flashbeats test fixes** (from #190) — loosen fbank GPU/CPU tolerance to 1e-3.
- **BEATs output fingerprints regenerated** for SDPA (torch_2_11_0 profile).

## What was dropped as redundant
- mil-ad part1 commits → already in #191.
- **#173 checkpoint-loading fix** → fully superseded by main's #177 (`init_config` + `load_packaged_yaml_mapping` already on main); the commit became empty.
- #190's EAT checksum refresh → already on main via #183.

## Validation (run locally on this branch)
- `test_official_beats_sdpa_equivalence` ✅ — SDPA numerically equivalent to legacy attention
- `test_batched_fbank` ✅, `test_beats_model` ✅, `test_official_models_checksums` ✅
- `test_official_models_regression` ✅ (31), `test_official_models_output_regression` ✅ (11) — beats fingerprints updated for SDPA, EAT/EffNet unchanged
- `test_api_integration` ✅

## Note
SDPA shifts exact output hashes (reduction-order), so beats fingerprints were regenerated for the `torch_2_11_0` band (the locked/CI torch). `torch_2_5_0`/`torch_2_6_0` beats values remain stale and would need regen on those torch versions if reintroduced.

Supersedes #108 and #190.

## Embeddings are preserved (SDPA is numerically equivalent)

The exact-SHA fingerprints for the 3 BEATs models changed, but the embeddings did **not** meaningfully change. Measured on a real pretrained BEATs checkpoint (SDPA vs the original manual attention, same weights, same audio):

```
output shape: (4, 96, 768)        value std ≈ 0.40, range [-6.82, 2.14]
max abs diff:   5.2e-06           ← largest single-element difference
mean abs diff:  3.8e-07
relative diff:  1.3e-05  (0.0013% of the signal's std)
cosine similarity (per-token):    min 0.99999976, mean 1.00000000
cosine similarity (clip-pooled):  0.99999994 … 1.0
```

The outputs are identical to ~7 significant figures (cosine sim ≈ 1.0 to 8 decimals). The max 5.2e-06 difference is float32 rounding noise from SDPA's fused kernel reducing in a different order than `softmax(QKᵀ)V` — downstream probes/retrieval/classification are unaffected.

**Why the SHA still changed:** SHA256 is bit-exact, so a single 1-ULP difference flips the whole hash. That's why the repo keeps two separate checks — `test_official_beats_sdpa_equivalence` (tolerance-based: atol 1e-5, e2e < 5e-4 — passes) proves numerical equivalence, while `test_official_models_output_regression` (exact SHA) catches any bit change. SDPA is a deliberate, numerically-equivalent kernel swap, so the exact fixtures are regenerated — same as any torch upgrade (hence the torch-version-keyed table).

_Caveat for reviewers:_ exact-SHA fingerprints are brittle for a fused-kernel attention path (hash can differ across GPU/CPU/torch builds even when correct). Consider switching the BEATs fingerprint check to a tolerance-based comparison.

### CODE REVIEW

File: avex/models/beats/backbone.py [hunk 1]
```diff
@@ -192,18 +191,15 @@ def extract_features(
...
             else:
-                # Normal layerdrop behavior
                 dropout_probability = np.random.random()
                 should_execute = not self.training or (dropout_probability > self.layerdrop)
 
             if should_execute:
                 x, z, pos_bias = layer(
                     x,
                     self_attn_padding_mask=padding_mask,
-                    need_weights=False,
```
Comment (GaganNarula) [inline avex/models/beats/backbone.py:206]: maybe add a comment here what this does or means

File: avex/models/beats/backbone.py [hunk 2]
```diff
@@ -238,7 +234,6 @@ def __init__(
         has_relative_attention_bias: bool = False,
         num_buckets: int = 0,
         max_distance: int = 0,
-        rescale_init: bool = False,
```
Comment (GaganNarula) [inline avex/models/beats/backbone.py:241]: also here

File: avex/models/beats/backbone.py (hunk 1 above)
Comment (nkundiushuti) [reply avex/models/beats/backbone.py:206]: Added in f4ae34d. This is the **LayerDrop** (stochastic depth) block: during training each encoder layer is randomly skipped with probability `self.layerdrop` for regularization; layers always run at inference, or when `disable_layerdrop=True` (used by the equivalence/fingerprint tests for determinism). The PR had stripped the inline comments — restored + expanded.

File: avex/models/beats/backbone.py (hunk 2 above)
Comment (nkundiushuti) [reply avex/models/beats/backbone.py:241]: Added a full constructor docstring in f4ae34d covering every param. Quick reference for the relative-position ones (BEATs-specific):

- `has_relative_attention_bias` — only layer 0 owns the T5-style relative-position-bias table; BEATs computes it once and threads it through later layers via `pos_bias`.
- `num_buckets` — number of buckets the relative offset (i−j) is mapped into.
- `max_distance` — distance at which bucketing saturates (farther pairs share the last bucket).
- `gru_rel_pos` — use BEATs' GRU-style *gated* relative position bias (per-query gate) instead of a plain additive bias.
- `deep_norm` / `encoder_layers` — DeepNorm residual scaling `(2·encoder_layers)**0.25` for stable deep-stack training.

Re the removed `rescale_init` (shown in this hunk): it only flipped `k_proj` bias off, was never set `True` for any BEATs model, and the new SDPA attention keeps `bias=True` on all projections — so the default (`k_bias=True`) behavior is unchanged. Removal is a no-op behaviorally, just dead-param cleanup.

File: avex/models/beats/backbone.py [hunk 3]
```diff
@@ -357,119 +372,66 @@ def forward(
...
-        else:
-            self.bias_k = self.bias_v = None
+        self.has_relative_attention_bias = has_relative_attention_bias
+        self.num_buckets = num_buckets
+        self.max_distance = max_distance
+        if self.has_relative_attention_bias:
+            self.relative_attention_bias = nn.Embedding(num_buckets, num_heads)
 
-        self.add_zero_attn = add_zero_attn
+        self.k_proj = nn.Linear(embed_dim, embed_dim, bias=bias)
```
Comment (GaganNarula) [inline avex/models/beats/backbone.py:414]: what was this quant_noise function ? why remove it ?

File: avex/models/beats/beats.py [hunk 4]
```diff
@@ -81,6 +211,14 @@ class BEATsConfig(BaseModel):
     max_distance: int = Field(800, description="Maximum distance for relative position embedding")
     gru_rel_pos: bool = Field(True, description="Apply gated relative position embedding")
 
+    # Spectrogram / preprocessing
+    sample_frequency: float = Field(16000.0, description="Audio sample rate in Hz")
+    num_mel_bins: int = Field(128, description="Number of mel filterbank bins")
+    frame_length: float = Field(25.0, description="Frame length in milliseconds")
+    frame_shift: float = Field(10.0, description="Frame shift (hop) in milliseconds")
+    fbank_mean: float = Field(15.41663, description="Mean for filterbank normalization")
```
Comment (GaganNarula) [inline avex/models/beats/beats.py:219]: so i guess these params are expected in BeatsConfig in esp-research right ?

File: avex/models/beats/beats.py [hunk 5]
```diff
@@ -154,43 +301,31 @@ def forward_padding_mask(
...
-                sample_frequency=16000,
-                frame_length=25,
-                frame_shift=10,
-            )
-            fbanks.append(fbank)
-        fbank = torch.stack(fbanks, dim=0)
-        fbank = (fbank - fbank_mean) / (2 * fbank_std)
-        return fbank
+        with torch.autocast(device_type=source.device.type, enabled=False):
+            fbank = self.fbank(source.float() * 2**15)
```
Comment (GaganNarula) [inline avex/models/beats/beats.py:322]: is this 2**15 from some repo ?

**Review by GaganNarula — APPROVED** [review]
Looks ok to me but i'm not an expert on the architecture.

File: avex/models/beats/backbone.py (hunk 3 above)
Comment (nkundiushuti) [reply avex/models/beats/backbone.py:414]: quant_noise was dropped as part of #160 "beats is fast" (the SDPA rewrite) same in David's flashbeats and
  in v2-dev. #192 just inherits it. it's a fairseq-era hook for iPQ quantization-aware training —
  quant_noise(m, p, block) returns m unchanged when p<=0, and BEATs always used q_noise=0 (never set higher anywhere). so it was a permanent no-op.

File: avex/models/beats/beats.py (hunk 4 above)
Comment (nkundiushuti) [reply avex/models/beats/beats.py:219]: I think esp-research already supplies them

File: avex/models/beats/beats.py (hunk 5 above)
Comment (nkundiushuti) [reply avex/models/beats/beats.py:322]: this was not introduced by #192.  it's in the original BEATs preprocess and on avex main's kaldi path (beats.py line 19). the batched fbank just keeps the same scaling so outputs match kaldi. 2**15 = 32768 = full-scale of 16-bit PCM. adio comes in as float [-1, 1] so kaldi/torchaudio fbank expects int16-range samples (±32768), so we rescale by 2^15.
